docs(sizing) resource sizing guidelines documentation (#1759)
Kong/docs.konghq.com: app/2.0.x/sizing-guidelines.md

Kong measures performance in two dimensions: _latency_ and _throughput_. _Latency_, in this context, refers to the delay between the downstream client sending a request and receiving a response. Kong measures latency introduced into the request in terms of microseconds or milliseconds. Increasing the number of Routes and/or Plugins in a Kong cluster will increase the amount of latency that is added into each request. _Throughput_ refers to number of simultaneous requests that Kong can process in a given time span, typically measured in seconds or minutes.

In general, these two dimensions have an inversely proportional relationship when all other factors remain the same: decreasing the latency introduced into each request allows the maximum throughput in Kong to increase, as there is less CPU time spent handling each request, and thus more CPU available for processing traffic as a whole. Kong is designed to scale horizontally to be able to add more overall compute power for configurations that add substantial latency into requests, while needing to meet specific throughput requirements.

> In general, Kong's maximum throughput is a CPU-bound dimension, and minumim latency is memory-bound. That is, adding more available compute power to the cluster to a latency-sensitive workload would be less beneficial than making available more memory for database caching. Likewise, throughput-sensitive workloads are dependant on both adequate memory and CPU resources, but adding more cache memory will do little to increase maximum throughput; adding more compute power by scaling Kong vertically and/or horizontally provides near-unlimited throughput capacity.

Reviewer: ```suggestion
In general, Kong's maximum throughput is a CPU-bound dimension, and the minimum latency is memory-bound. That is, adding more available compute power to the cluster to a latency-sensitive workload would be less beneficial than making available more memory for database caching. Likewise, throughput-sensitive workloads are dependant on both adequate memory and CPU resources, but adding more cache memory will do little to increase maximum throughput. Adding more compute power by scaling Kong vertically and/or horizontally provides near-unlimited throughput capacity.
```

> Performance benchmarking and optimization as a whole is a complex exercise that must account for a variety of factors, including those external to Kong, such as the behavior of upstream services, or the health of the underlying hardware on which Kong is running.

Reviewer: ```suggestion
Performance benchmarking and optimization as a whole is a complex exercise that must account for a variety of factors. These factors can be external to Kong, such as the behavior of upstream services, or the health of the underlying hardware on which Kong is running.
```

## Performance Characteristics

> There are a number of things that impact how Kong behaves, including:

Reviewer: ```suggestion
There are a number of factors that impact Kong's behavior, including:
```

* **Number of configured Routes and Services**. Increasing the count of Routes and Services on the cluster will require more CPU to 

> * **Number of configured Consumers and Credentials**. Consumer and credential data is stored in Kong's datastore (either Postgres or Cassandra, or the `kong.yml` file in dbless environments); Kong caches this data in memory to reduce database load and latency during request processing. Increase the count of Consumers and Credentials will require more memory available for Kong to hold data in cache.

Reviewer: ```suggestion
* **Number of configured Consumers and Credentials**. Consumer and credential data rests in Kong's datastore (either PostgreSQL or Cassandra, or the `kong.yml` file in DB-less environments). Kong caches this data in memory to reduce the database load and latency during request processing. To increase the count of Consumers and Credentials requires more memory available for Kong to hold data in a cache.
```

* **Number of configured Plugins**.  Increasing the count of Plugins on the cluster will require more CPU to iterate through plugins during request processing. Executing plugins comes with a varying cost depending on the nature of the plugin; for example, a lightweight authentication plugin like `key-auth` requires less resource availability than a plugin that performs complex transformations of an HTTP request or response.

> * **Cardinality of configured Plugins**. This refers to the number of distinct plugin types that are configured on the cluster. With each additional plugin type added to the cluster, Kong spends more time evaluating whether to execute a given plugin for a given request. Increasing the cardinality of configured plugins requires more CPU power, as the process to evaluate plugins is a CPU-constrained task.

Reviewer: ```suggestion
* **Cardinality of configured Plugins**. _Cardinality_ refers to the number of distinct plugin types that are configured on the cluster. With each additional plugin _type_ added to the cluster, Kong spends more time evaluating whether to execute a given plugin for a given request. Increasing the cardinality of configured plugins requires more CPU power, as the process to evaluate plugins is a CPU-constrained task.
```

> * **Request and response size**. Requests with large HTTP bodies, either in the request or response, will take longer to process, as Kong must buffer the request to disk before proxying it. This allows Kong to handle a large volume of simultaneous traffic without running out of memory, but the nature of buffered requests can result in increased latency.

Reviewer: ```suggestion
* **Request and response size**. Requests with large HTTP bodies, either in the request or response, will take longer to process, as Kong must buffer the request to disk before proxying it. This process allows Kong to handle a large volume of simultaneous traffic without running out of memory, but the nature of buffered requests can result in increased latency.
```
